README.md
LPegLJ v1.0
=============

LPeg Parser in pure LuaJIT  
(straight Lua + FFI translation of LPeg C code)   
based on LPeg v1.0 - PEG pattern matching for Lua  
Lua.org & PUC-Rio  written by Roberto Ierusalimschy  
http://www.inf.puc-rio.br/~roberto/lpeg/

left recursion support is based on Sérgio Medeiros algorithm
http://arxiv.org/abs/1207.0443

### Usage:  
```Lua
local lpeglj = require"lpeglj"  
local pattern = lpeglj.P("a") 
-- then:
lpeglj.match(pattern, "a") 
-- or, equivalently:  
pattern:match("a")  
```

### Compatibility:

- full syntactical and functional backward compatibility with LPeg v1.0
- works only with LuaJIT 2.x  

### Differences from LPeg v1.0:

Description in doc/USAGE.md

- LPegLJ supports direct and indirect left recursion based on Sérgio Medeiros algorithm (http://arxiv.org/abs/1207.0443)
- patterns can be saved and loaded
- supports memoization (restricted) - useful for complex grammars
- can be used in stream mode (infinite parsing)
- VM action runtime listing (tracing) for debugging purposes


CHANGELOG.md
1.0.0.0 : September 30, 2015
- include changes and bug fixes from LPeg v1.0 
- added VM runtime listing (tracing) for debugging purposes

0.12.2 : July 10, 2014

- added restricted memoization
- stream support (infinite parsing)

0.12.1 : December 30, 2013

- speed improvement
- support direct and indirect left recursion based on Sérgio Medeiros algorithm (http://arxiv.org/abs/1207.0443)
- loading and saving patterns

0.12 : July 14, 2013: Initial release

- LPeg Parser in pure LuaJIT based on LPeg v.12

tests/testlr.lua
local lpeg = require"lpeglj"
local re = require"re"

local m = lpeg

local function checkeq(x, y, p)
    if p then print(x, y) end
    if type(x) ~= "table" then assert(x == y)
    else
        for k, v in pairs(x) do checkeq(v, y[k], p) end
        for k, v in pairs(y) do checkeq(v, x[k], p) end
    end
end

print"Tests for LPegLJ left recursion"

assert(type(m.version()) == "string")
print("version " .. m.version())

m.enableleftrecursion(true)

--[[
direct left recursion
E ← E + n / n
--]]

local pat = m.P{
    "E";
    E = m.V"E" * '+' * "n" + "n",
}

assert(pat:match("n+n+n") == 6)

--[[
indirect left recursion
L ← P.x / x
P ← P(n) / L
--]]

local pat = m.P{
    "L";
    L = m.V"P" * ".x" + "x",
    P = m.V"P" * "(n)" + m.V"L"
}

assert(pat:match("x(n)(n).x(n).x") == 15)

--[[
left and right recursion with precedence rules
E ← E1 + E2 / E1 − E2 / E2 ∗ E3 / E2 ÷ E3 / E3 ∗∗ E3 / − E4 / (E1) / n
--]]


local pat = m.P{
    "E",
    E = m.V("E", 1) * m.S'+-' * m.V("E", 2) +
            m.V("E", 2) * m.S'*/' * m.V("E", 3) +
            m.V("E", 3) * '**' * m.V("E", 3) +
            '-' * m.V("E", 4) +
            '(' * m.V("E") * ')' +
            m.R'09' ^ 1,
}

assert(pat:match("-1*(6+2/4+3-1)**2") == 18)

--[[
left and right recursion with precedence rules
E ← E1 + E2 / E1 − E2 / E2 ∗ E3 / E2 ÷ E3 / E3 ∗∗ E3 / − E4 / (E1) / n
create AST tree
--]]


local pat = m.P{
    "E",
    E = m.Ct(m.V("E", 1) * m.C(m.S'+-') * m.V("E", 2) +
            m.V("E", 2) * m.C(m.S'*/') * m.V("E", 3) +
            m.V("E", 3) * m.C('**') * m.V("E", 3) +
            m.C('-') * m.V("E", 4) +
            '(' * m.V("E") * ')' +
            m.C(m.R'09' ^ 1)),
}

local ASTtree = pat:match("1+1+1")
checkeq(ASTtree, { { { "1" }, "+", { "1" } }, "+", { "1" } })

local ASTtree = pat:match("-1*(6+2/4+3-1)**2")
checkeq(ASTtree, { { "-", { "1" } }, "*", { { { { { { "6" }, "+", { { "2" }, "/", { "4" } } }, "+", { "3" } }, "-", { "1" } } }, "**", { "2" } } })

-- using re module with precedence (the same example as above)
-- call_nonterminal : precedence_level or <call_nonterminal : precedence_level >

local pat = [[
     E <- (E:1 {[+-]} E:2 /
          E:2 {[*/]} E:3 /
          E:3 {'**'} E:3 /
          {'-'} E:4 /
          '(' E ')' /
          {[0-9]+}) -> {}
]]

local ASTtree = re.match("-1*(6+2/4+3-1)**2", pat)
checkeq(ASTtree, { { "-", { "1" } }, "*", { { { { { { "6" }, "+", { { "2" }, "/", { "4" } } }, "+", { "3" } }, "-", { "1" } } }, "**", { "2" } } })

--[[
simple evaluator
E ← E1 + E2 / E1 − E2 / E2 ∗ E3 / E2 ÷ E3 / E3 ∗∗ E3 / − E4 / (E1) / n
--]]

local eval = function(s, i, p1, p2, p3)
    local res
    if p2 == '+' then
        res = p1 + p3
    elseif p2 == '-' then
        res = p1 - p3
    elseif p2 == '*' then
        res = p1 * p3
    elseif p2 == '/' then
        res = p1 / p3
    elseif p1 == '-' then
        res = -p2
    elseif p2 == '**' then
        res = p1 ^ p3
    else
        res = p1
    end
    return true, res
end


local pat = m.P{
    "E",
    E = m.Cmt(m.V("E", 1) * m.C(m.S'+-') * m.V("E", 2) +
            m.V("E", 2) * m.C(m.S'*/') * m.V("E", 3) +
            m.V("E", 3) * m.C('**') * m.V("E", 3) +
            m.C('-') * m.V("E", 4) +
            '(' * m.V("E") * ')' +
            m.C(m.R'09' ^ 1), eval),
}

assert(pat:match("-1*(6+2/4+3-1)**2") == -72.25)


local pat = m.P{
    "E",
    E = m.V("E", 1) * '+' * m.V("E", 2) / function(c1, c2) return c1 + c2 end +
            m.V("E", 1) * '-' * m.V("E", 2) / function(c1, c2) return c1 - c2 end +
            m.V("E", 2) * '*' * m.V("E", 3) / function(c1, c2) return c1 * c2 end +
            m.V("E", 2) * '/' * m.V("E", 3) / function(c1, c2) return c1 / c2 end +
            m.V("E", 3) * '**' * m.V("E", 3) / function(c1, c2) return c1 ^ c2 end +
            '-' * m.V("E", 4) / function(c1) return -c1 end +
            '(' * m.V("E") * ')' +
            m.C(m.R'09' ^ 1),
}

assert(pat:match("-1*(6+2/4+3-1)**2") == -72.25)

local def = {
    plus = function(p1, p2) return p1 + p2 end,
    minus = function(p1, p2) return p1 - p2 end,
    mult = function(p1, p2) return p1 * p2 end,
    div = function(p1, p2) return p1 / p2 end,
    pow = function(p1, p2) return p1 ^ p2 end,
    uminus = function(p1) return -p1 end,
    errfce = function(o, i)
        local errstr = o .. '\n' .. (' '):rep(i) .. '^' .. '\n'
        io.write(errstr)
        return false
    end,
}

local pat = [[
     P <-  E s (!. / error)
     s <- %s*
     error <- '' => errfce
     E <- (E:1 s'+' E:2) -> plus /
          (E:1 s'-' E:2) -> minus /
          (E:2 s'*' E:3) -> mult /
          (E:2 s'/' E:3) -> div /
          (E:3 s'**' E:3)-> pow /
          (s'-' E:4) -> uminus /
          s'(' E s')' /
          s{[0-9]+} /
          error
]]

local pat = re.compile(pat, def)
assert(re.match("-1 * (6 + 2 / 4 + 3 - 1)**2", pat) == -72.25)

print"OK"


src/re.lua
-- $Id: re.lua,v 1.44 2013/03/26 20:11:40 roberto Exp $

-- imported functions and modules
local tonumber, type, print, error = tonumber, type, print, error
local setmetatable = setmetatable
local m = require"lpeglj"

-- 'm' will be used to parse expressions, and 'mm' will be used to
-- create expressions; that is, 're' runs on 'm', creating patterns
-- on 'mm'
local mm = m

-- pattern's metatable
local mt = getmetatable(mm.P(0))
mt = m.version() == "0.12.2LJ" and m or mt



-- No more global accesses after this point
local version = _VERSION
if version == "Lua 5.2" then _ENV = nil end


local any = m.P(1)


-- Pre-defined names
local Predef = { nl = m.P"\n" }


local mem
local fmem
local gmem


local function updatelocale ()
  mm.locale(Predef)
  Predef.a = Predef.alpha
  Predef.c = Predef.cntrl
  Predef.d = Predef.digit
  Predef.g = Predef.graph
  Predef.l = Predef.lower
  Predef.p = Predef.punct
  Predef.s = Predef.space
  Predef.u = Predef.upper
  Predef.w = Predef.alnum
  Predef.x = Predef.xdigit
  Predef.A = any - Predef.a
  Predef.C = any - Predef.c
  Predef.D = any - Predef.d
  Predef.G = any - Predef.g
  Predef.L = any - Predef.l
  Predef.P = any - Predef.p
  Predef.S = any - Predef.s
  Predef.U = any - Predef.u
  Predef.W = any - Predef.w
  Predef.X = any - Predef.x
  mem = {}    -- restart memoization
  fmem = {}
  gmem = {}
  local mt = {__mode = "v"}
  setmetatable(mem, mt)
  setmetatable(fmem, mt)
  setmetatable(gmem, mt)
end


updatelocale()



local I = m.P(function (s,i) print(i, s:sub(1, i-1)); return i end)


local function getdef (id, defs)
  local c = defs and defs[id]
  if not c then error("undefined name: " .. id) end
  return c
end


local function patt_error (s, i)
  local msg = (#s < i + 20) and s:sub(i)
                             or s:sub(i,i+20) .. "..."
  msg = ("pattern error near '%s'"):format(msg)
  error(msg, 2)
end

local function mult (p, n)
  local np = mm.P(true)
  while n >= 1 do
    if n%2 >= 1 then np = np * p end
    p = p * p
    n = n/2
  end
  return np
end

local function equalcap (s, i, c)
  if type(c) ~= "string" then return nil end
  local e = #c + i
  if type(s) == 'function' then  -- stream mode
      if s(i, e - 1) == c then return e else return nil end
  else
      if s:sub(i, e - 1) == c then return e else return nil end
  end
end


local S = (Predef.space + "--" * (any - Predef.nl)^0)^0

local name = m.R("AZ", "az", "__") * m.R("AZ", "az", "__", "09")^0

local arrow = S * "<-"

local seq_follow = m.P"/" + ")" + "}" + ":}" + "~}" + "|}" + (name * arrow) + -1

name = m.C(name)


-- a defined name only have meaning in a given environment
local Def = name * m.Carg(1)

local num = m.C(m.R"09"^1) * S / tonumber

local String = "'" * m.C((any - "'")^0) * "'" +
               '"' * m.C((any - '"')^0) * '"'


local defined = "%" * Def / function (c,Defs)
  local cat =  Defs and Defs[c] or Predef[c]
  if not cat then error ("name '" .. c .. "' undefined") end
  return cat
end

local Range = m.Cs(any * (m.P"-"/"") * (any - "]")) / mm.R

local item = defined + Range + m.C(any)

local Class =
    "["
  * (m.C(m.P"^"^-1))    -- optional complement symbol
  * m.Cf(item * (item - "]")^0, mt.__add) /
                          function (c, p) return c == "^" and any - p or p end
  * "]"

local function adddef (t, k, exp)
  if t[k] then
    error("'"..k.."' already defined as a rule")
  else
    t[k] = exp
  end
  return t
end

local function firstdef (n, r) return adddef({n}, n, r) end


local function NT (n, b, p)
  if not b then
    error("rule '"..n.."' used outside a grammar")
  else return mm.V(n, p or 0)
  end
end


local exp = m.P{ "Exp",
  Exp = S * ( m.V"Grammar"
            + m.Cf(m.V"Seq" * ("/" * S * m.V"Seq")^0, mt.__add) );
  Seq = m.Cf(m.Cc(m.P"") * m.V"Prefix"^0 , mt.__mul)
        * (#seq_follow + patt_error);
  Prefix = "&" * S * m.V"Prefix" / mt.__len
         + "!" * S * m.V"Prefix" / mt.__unm
         + m.V"Suffix";
  Suffix = m.Cf(m.V"Primary" * S *
          ( ( m.P"+" * m.Cc(1, mt.__pow)
            + m.P"*" * m.Cc(0, mt.__pow)
            + m.P"?" * m.Cc(-1, mt.__pow)
            + "^" * ( m.Cg(num * m.Cc(mult))
                    + m.Cg(m.C(m.S"+-" * m.R"09"^1) * m.Cc(mt.__pow))
                    )
            + "->" * S * ( m.Cg((String + num) * m.Cc(mt.__div))
                         + m.P"{}" * m.Cc(nil, m.Ct)
                         + m.Cg(Def / getdef * m.Cc(mt.__div))
                         )
            + "=>" * S * m.Cg(Def / getdef * m.Cc(m.Cmt))
            ) * S
          )^0, function (a,b,f) return f(a,b) end );
  Primary = "(" * m.V"Exp" * ")"
            + String / mm.P
            + Class
            + defined
            + "{:" * (name * ":" + m.Cc(nil)) * m.V"Exp" * ":}" /
                     function (n, p) return mm.Cg(p, n) end
            + "=" * name / function (n) return mm.Cmt(mm.Cb(n), equalcap) end
            + m.P"{}" / mm.Cp
            + "{~" * m.V"Exp" * "~}" / mm.Cs
            + "{|" * m.V"Exp" * "|}" / mm.Ct
            + "{" * m.V"Exp" * "}" / mm.C
            + m.P"." * m.Cc(any)
            + (name * m.Cb("G") * (S * ":" * S * num)^-1 * -arrow + "<" * name * m.Cb("G") * (S * ":" * S * num)^-1 * ">") / NT;
  Definition = name * arrow * m.V"Exp";
  Grammar = m.Cg(m.Cc(true), "G") *
            m.Cf(m.V"Definition" / firstdef * m.Cg(m.V"Definition")^0,
              adddef) / mm.P
}

local pattern = S * m.Cg(m.Cc(false), "G") * exp / mm.P * (-any + patt_error)


local function compile (p, defs)
  if mm.type(p) == "pattern" then return p end   -- already compiled
  local cp = pattern:match(p, 1, defs)
  if not cp then error("incorrect pattern", 3) end
  return cp
end

local function match (s, p, i)
  local cp = mem[p]
  if not cp then
    cp = compile(p)
    mem[p] = cp
  end
  return cp:match(s, i or 1)
end

local function streammatch (p, i)
    local cp = mem[p]
    if not cp then
        cp = compile(p)
        mem[p] = cp
    end
    return cp:streammatch(i or 1)
end

local function emulatestreammatch(s, p, i)
    local cp = mem[p]
    if not cp then
        cp = compile(p)
        mem[p] = cp
    end
    return cp:emulatestreammatch(s, i or 1)
end

local function find (s, p, i)
  local cp = fmem[p]
  if not cp then
    cp = compile(p) / 0
    cp = mm.P{ mm.Cp() * cp * mm.Cp() + 1 * mm.V(1) }
    fmem[p] = cp
  end
  local i, e = cp:match(s, i or 1)
  if i then return i, e - 1
  else return i
  end
end

local function gsub (s, p, rep)
  local g = gmem[p] or {}   -- ensure gmem[p] is not collected while here
  gmem[p] = g
  local cp = g[rep]
  if not cp then
    cp = compile(p)
    cp = mm.Cs((cp / rep + 1)^0)
    g[rep] = cp
  end
  return cp:match(s)
end


-- exported names
local re = {
  compile = compile,
  match = match,
  streammatch = streammatch,
  emulatestreammatch = emulatestreammatch,
  find = find,
  gsub = gsub,
  updatelocale = updatelocale,
}

if version == "Lua 5.1" then _G.re = re end

return re


src/streamre.lua
-- $Id: re.lua,v 1.44 2013/03/26 20:11:40 roberto Exp $

-- imported functions and modules
local tonumber, type, print, error = tonumber, type, print, error
local setmetatable = setmetatable
local m = require"lpeglj"
m.match = m.emulatestreammatch  -- change standard match with stream emulation

-- 'm' will be used to parse expressions, and 'mm' will be used to
-- create expressions; that is, 're' runs on 'm', creating patterns
-- on 'mm'
local mm = m

-- pattern's metatable
local mt = getmetatable(mm.P(0))
mt = m.version() == "0.12.2LJ" and m or mt



-- No more global accesses after this point
local version = _VERSION
if version == "Lua 5.2" then _ENV = nil end


local any = m.P(1)


-- Pre-defined names
local Predef = { nl = m.P"\n" }


local mem
local fmem
local gmem


local function updatelocale ()
  mm.locale(Predef)
  Predef.a = Predef.alpha
  Predef.c = Predef.cntrl
  Predef.d = Predef.digit
  Predef.g = Predef.graph
  Predef.l = Predef.lower
  Predef.p = Predef.punct
  Predef.s = Predef.space
  Predef.u = Predef.upper
  Predef.w = Predef.alnum
  Predef.x = Predef.xdigit
  Predef.A = any - Predef.a
  Predef.C = any - Predef.c
  Predef.D = any - Predef.d
  Predef.G = any - Predef.g
  Predef.L = any - Predef.l
  Predef.P = any - Predef.p
  Predef.S = any - Predef.s
  Predef.U = any - Predef.u
  Predef.W = any - Predef.w
  Predef.X = any - Predef.x
  mem = {}    -- restart memoization
  fmem = {}
  gmem = {}
  local mt = {__mode = "v"}
  setmetatable(mem, mt)
  setmetatable(fmem, mt)
  setmetatable(gmem, mt)
end


updatelocale()



local I = m.P(function (s,i) print(i, s:sub(1, i-1)); return i end)


local function getdef (id, defs)
  local c = defs and defs[id]
  if not c then error("undefined name: " .. id) end
  return c
end


local function patt_error (s, i)
  local msg = (#s(1,-1) < i + 20) and s(1,-1):sub(i)
                             or s(1,-1):sub(i,i+20) .. "..."
  msg = ("pattern error near '%s'"):format(msg)
  error(msg, 2)
end

local function mult (p, n)
  local np = mm.P(true)
  while n >= 1 do
    if n%2 >= 1 then np = np * p end
    p = p * p
    n = n/2
  end
  return np
end

local function equalcap (s, i, c)
  if type(c) ~= "string" then return nil end
  local e = #c + i
  if s(1,-1):sub(i, e - 1) == c then return e else return nil end
end


local S = (Predef.space + "--" * (any - Predef.nl)^0)^0

local name = m.R("AZ", "az", "__") * m.R("AZ", "az", "__", "09")^0

local arrow = S * "<-"

local seq_follow = m.P"/" + ")" + "}" + ":}" + "~}" + "|}" + (name * arrow) + -1

name = m.C(name)


-- a defined name only have meaning in a given environment
local Def = name * m.Carg(1)

local num = m.C(m.R"09"^1) * S / tonumber

local String = "'" * m.C((any - "'")^0) * "'" +
               '"' * m.C((any - '"')^0) * '"'


local defined = "%" * Def / function (c,Defs)
  local cat =  Defs and Defs[c] or Predef[c]
  if not cat then error ("name '" .. c .. "' undefined") end
  return cat
end

local Range = m.Cs(any * (m.P"-"/"") * (any - "]")) / mm.R

local item = defined + Range + m.C(any)

local Class =
    "["
  * (m.C(m.P"^"^-1))    -- optional complement symbol
  * m.Cf(item * (item - "]")^0, mt.__add) /
                          function (c, p) return c == "^" and any - p or p end
  * "]"

local function adddef (t, k, exp)
  if t[k] then
    error("'"..k.."' already defined as a rule")
  else
    t[k] = exp
  end
  return t
end

local function firstdef (n, r) return adddef({n}, n, r) end


local function NT (n, b, p)
  if not b then
    error("rule '"..n.."' used outside a grammar")
  else return mm.V(n, p or 0)
  end
end


local exp = m.P{ "Exp",
  Exp = S * ( m.V"Grammar"
            + m.Cf(m.V"Seq" * ("/" * S * m.V"Seq")^0, mt.__add) );
  Seq = m.Cf(m.Cc(m.P"") * m.V"Prefix"^0 , mt.__mul)
        * (#seq_follow + patt_error);
  Prefix = "&" * S * m.V"Prefix" / mt.__len
         + "!" * S * m.V"Prefix" / mt.__unm
         + m.V"Suffix";
  Suffix = m.Cf(m.V"Primary" * S *
          ( ( m.P"+" * m.Cc(1, mt.__pow)
            + m.P"*" * m.Cc(0, mt.__pow)
            + m.P"?" * m.Cc(-1, mt.__pow)
            + "^" * ( m.Cg(num * m.Cc(mult))
                    + m.Cg(m.C(m.S"+-" * m.R"09"^1) * m.Cc(mt.__pow))
                    )
            + "->" * S * ( m.Cg((String + num) * m.Cc(mt.__div))
                         + m.P"{}" * m.Cc(nil, m.Ct)
                         + m.Cg(Def / getdef * m.Cc(mt.__div))
                         )
            + "=>" * S * m.Cg(Def / getdef * m.Cc(m.Cmt))
            ) * S
          )^0, function (a,b,f) return f(a,b) end );
  Primary = "(" * m.V"Exp" * ")"
            + String / mm.P
            + Class
            + defined
            + "{:" * (name * ":" + m.Cc(nil)) * m.V"Exp" * ":}" /
                     function (n, p) return mm.Cg(p, n) end
            + "=" * name / function (n) return mm.Cmt(mm.Cb(n), equalcap) end
            + m.P"{}" / mm.Cp
            + "{~" * m.V"Exp" * "~}" / mm.Cs
            + "{|" * m.V"Exp" * "|}" / mm.Ct
            + "{" * m.V"Exp" * "}" / mm.C
            + m.P"." * m.Cc(any)
            + (name * m.Cb("G") * (S * ":" * S * num)^-1 * -arrow + "<" * name * m.Cb("G") * (S * ":" * S * num)^-1 * ">") / NT;
  Definition = name * arrow * m.V"Exp";
  Grammar = m.Cg(m.Cc(true), "G") *
            m.Cf(m.V"Definition" / firstdef * m.Cg(m.V"Definition")^0,
              adddef) / mm.P
}

local pattern = S * m.Cg(m.Cc(false), "G") * exp / mm.P * (-any + patt_error)


local function compile (p, defs)
  if mm.type(p) == "pattern" then return p end   -- already compiled
  local cp = pattern:match(p, 1, defs)
  if not cp then error("incorrect pattern", 3) end
  return cp
end

local function match (s, p, i)
  local cp = mem[p]
  if not cp then
    cp = compile(p)
    mem[p] = cp
  end
  return cp:match(s, i or 1)
end

local function find (s, p, i)
  local cp = fmem[p]
  if not cp then
    cp = compile(p) / 0
    cp = mm.P{ mm.Cp() * cp * mm.Cp() + 1 * mm.V(1) }
    fmem[p] = cp
  end
  local i, e = cp:match(s, i or 1)
  if i then return i, e - 1
  else return i
  end
end

local function gsub (s, p, rep)
  local g = gmem[p] or {}   -- ensure gmem[p] is not collected while here
  gmem[p] = g
  local cp = g[rep]
  if not cp then
    cp = compile(p)
    cp = mm.Cs((cp / rep + 1)^0)
    g[rep] = cp
  end
  return cp:match(s)
end


-- exported names
local re = {
  compile = compile,
  match = match,
  find = find,
  gsub = gsub,
  updatelocale = updatelocale,
}

if version == "Lua 5.1" then _G.re = re end

return re


src/lpprint.lua
--[[
LPEGLJ
lpprint.lua
Tree, code and debug print function (only for debuging)
Copyright (C) 2014 Rostislav Sacek.
based on LPeg v0.12 - PEG pattern matching for Lua
Lua.org & PUC-Rio  written by Roberto Ierusalimschy
http://www.inf.puc-rio.br/~roberto/lpeg/

** Permission is hereby granted, free of charge, to any person obtaining
** a copy of this software and associated documentation files (the
** "Software"), to deal in the Software without restriction, including
** without limitation the rights to use, copy, modify, merge, publish,
** distribute, sublicense, and/or sell copies of the Software, and to
** permit persons to whom the Software is furnished to do so, subject to
** the following conditions:
**
** The above copyright notice and this permission notice shall be
** included in all copies or substantial portions of the Software.
**
** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
** EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
** MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
** IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
** CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
** TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
** SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
**
** [ MIT license: http://www.opensource.org/licenses/mit-license.php ]
--]]

local ffi = require"ffi"
local band, rshift, lshift = bit.band, bit.rshift, bit.lshift

ffi.cdef[[
  int isprint ( int c );
]]

local RuleLR = 0x10000
local Ruleused = 0x20000

-- {======================================================
-- Printing patterns (for debugging)
-- =======================================================

local TChar = 0
local TSet = 1
local TAny = 2 -- standard PEG elements
local TTrue = 3
local TFalse = 4
local TRep = 5
local TSeq = 6
local TChoice = 7
local TNot = 8
local TAnd = 9
local TCall = 10
local TOpenCall = 11
local TRule = 12 -- sib1 is rule's pattern, sib2 is 'next' rule
local TGrammar = 13 -- sib1 is initial (and first) rule
local TBehind = 14 -- match behind
local TCapture = 15 -- regular capture
local TRunTime = 16 -- run-time capture

local IAny = 0 -- if no char, fail
local IChar = 1 -- if char != aux, fail
local ISet = 2 -- if char not in val, fail
local ITestAny = 3 -- in no char, jump to 'offset'
local ITestChar = 4 -- if char != aux, jump to 'offset'
local ITestSet = 5 -- if char not in val, jump to 'offset'
local ISpan = 6 -- read a span of chars in val
local IBehind = 7 -- walk back 'aux' characters (fail if not possible)
local IRet = 8 -- return from a rule
local IEnd = 9 -- end of pattern
local IChoice = 10 -- stack a choice; next fail will jump to 'offset'
local IJmp = 11 -- jump to 'offset'
local ICall = 12 -- call rule at 'offset'
local IOpenCall = 13 -- call rule number 'offset' (must be closed to a ICall)
local ICommit = 14 -- pop choice and jump to 'offset'
local IPartialCommit = 15 -- update top choice to current position and jump
local IBackCommit = 16 -- "fails" but jump to its own 'offset'
local IFailTwice = 17 -- pop one choice and then fail
local IFail = 18 -- go back to saved state on choice and jump to saved offset
local IGiveup = 19 -- internal use
local IFullCapture = 20 -- complete capture of last 'off' chars
local IOpenCapture = 21 -- start a capture
local ICloseCapture = 22
local ICloseRunTime = 23

local Cclose = 0
local Cposition = 1
local Cconst = 2
local Cbackref = 3
local Carg = 4
local Csimple = 5
local Ctable = 6
local Cfunction = 7
local Cquery = 8
local Cstring = 9
local Cnum = 10
local Csubst = 11
local Cfold = 12
local Cruntime = 13
local Cgroup = 14


-- number of siblings for each tree
local numsiblings = {
    [TRep] = 1,
    [TSeq] = 2,
    [TChoice] = 2,
    [TNot] = 1,
    [TAnd] = 1,
    [TRule] = 2,
    [TGrammar] = 1,
    [TBehind] = 1,
    [TCapture] = 1,
    [TRunTime] = 1,
}
local names = {
    [IAny] = "any",
    [IChar] = "char",
    [ISet] = "set",
    [ITestAny] = "testany",
    [ITestChar] = "testchar",
    [ITestSet] = "testset",
    [ISpan] = "span",
    [IBehind] = "behind",
    [IRet] = "ret",
    [IEnd] = "end",
    [IChoice] = "choice",
    [IJmp] = "jmp",
    [ICall] = "call",
    [IOpenCall] = "open_call",
    [ICommit] = "commit",
    [IPartialCommit] = "partial_commit",
    [IBackCommit] = "back_commit",
    [IFailTwice] = "failtwice",
    [IFail] = "fail",
    [IGiveup] = "giveup",
    [IFullCapture] = "fullcapture",
    [IOpenCapture] = "opencapture",
    [ICloseCapture] = "closecapture",
    [ICloseRunTime] = "closeruntime"
}

local function printcharset(st)
    io.write("[");
    local i = 0
    while i <= 255 do
        local first = i;
        while band(st[rshift(i, 5)], lshift(1, band(i, 31))) ~= 0 and i <= 255 do
            i = i + 1
        end
        if i - 1 == first then -- unary range?
            io.write(("(%02x)"):format(first))
        elseif i - 1 > first then -- non-empty range?
            io.write(("(%02x-%02x)"):format(first, i - 1))
        end
        i = i + 1
    end
    io.write("]")
end

local modes = {
    [Cclose] = "close",
    [Cposition] = "position",
    [Cconst] = "constant",
    [Cbackref] = "backref",
    [Carg] = "argument",
    [Csimple] = "simple",
    [Ctable] = "table",
    [Cfunction] = "function",
    [Cquery] = "query",
    [Cstring] = "string",
    [Cnum] = "num",
    [Csubst] = "substitution",
    [Cfold] = "fold",
    [Cruntime] = "runtime",
    [Cgroup] = "group"
}

local function printcapkind(kind)
    io.write(("%s"):format(modes[kind]))
end

local function printjmp(p, index)
    io.write(("-> %d"):format(index + p[index].offset))
end


local function printinst(p, index, valuetable)
    local code = p[index].code
    io.write(("%02d: %s "):format(index, names[code]))
    if code == IChar then
        io.write(("'%s'"):format(string.char(p[index].val)))
    elseif code == ITestChar then
        io.write(("'%s'"):format(string.char(p[index].val)))
        printjmp(p, index)
    elseif code == IFullCapture then
        printcapkind(band(p[index].val, 0x0f));
        io.write((" (size = %d)  (idx = %s)"):format(band(rshift(p[index].val, 4), 0xF), tostring(valuetable[p[index].offset])))
    elseif code == IOpenCapture then
        printcapkind(band(p[index].val, 0x0f))
        io.write((" (idx = %s)"):format(tostring(valuetable[p[index].offset])))
    elseif code == ISet then
        printcharset(valuetable[p[index].val]);
    elseif code == ITestSet then
        printcharset(valuetable[p[index].val])
        printjmp(p, index);
    elseif code == ISpan then
        printcharset(valuetable[p[index].val]);
    elseif code == IOpenCall then
        io.write(("-> %d"):format(p[index].offset))
    elseif code == IBehind then
        io.write(("%d"):format(p[index].val))
    elseif code == IJmp or code == ICall or code == ICommit or code == IChoice or
            code == IPartialCommit or code == IBackCommit or code == ITestAny then
        printjmp(p, index);
        if (code == ICall or code == IJmp) and p[index].aux > 0 then
            io.write(' ', valuetable[p[index].aux])
        end
    end
    io.write("\n")
end


local function printpatt(p, valuetable)
    for i = 0, p.size - 1 do
        printinst(p.p, i, valuetable);
    end
end


local function printcap(cap, index, valuetable)
    printcapkind(cap[index].kind)
    io.write((" (idx: %s - size: %d) -> %d\n"):format(valuetable[cap[index].idx], cap[index].siz, cap[index].s))
end


local function printcaplist(cap, limit, valuetable)
    io.write(">======\n")
    local index = 0
    while cap[index].s and index < limit do
        printcap(cap, index, valuetable)
        index = index + 1
    end
    io.write("=======\n")
end

-- ======================================================



-- {======================================================
-- Printing trees (for debugging)
-- =======================================================

local tagnames = {
    [TChar] = "char",
    [TSet] = "set",
    [TAny] = "any",
    [TTrue] = "true",
    [TFalse] = "false",
    [TRep] = "rep",
    [TSeq] = "seq",
    [TChoice] = "choice",
    [TNot] = "not",
    [TAnd] = "and",
    [TCall] = "call",
    [TOpenCall] = "opencall",
    [TRule] = "rule",
    [TGrammar] = "grammar",
    [TBehind] = "behind",
    [TCapture] = "capture",
    [TRunTime] = "run-time"
}


local function printtree(tree, ident, index, valuetable)
    for i = 1, ident do
        io.write(" ")
    end
    local tag = tree[index].tag
    io.write(("%s"):format(tagnames[tag]))
    if tag == TChar then
        local c = tree[index].val
        if ffi.C.isprint(c) then
            io.write((" '%c'\n"):format(c))
        else
            io.write((" (%02X)\n"):format(c))
        end
    elseif tag == TSet then
        printcharset(valuetable[tree[index].val]);
        io.write("\n")
    elseif tag == TOpenCall or tag == TCall then
        io.write((" key: %s\n"):format(tostring(valuetable[tree[index].val])))
    elseif tag == TBehind then
        io.write((" %d\n"):format(tree[index].val))
        printtree(tree, ident + 2, index + 1, valuetable);
    elseif tag == TCapture then
        io.write((" cap: %s   n: %s\n"):format(modes[bit.band(tree[index].cap, 0xffff)], valuetable[tree[index].val]))
        printtree(tree, ident + 2, index + 1, valuetable);
    elseif tag == TRule then
        local extra = bit.band(tree[index].cap, RuleLR) == RuleLR and ' left recursive' or ''
        extra = extra .. (bit.band(tree[index].cap, Ruleused) ~= Ruleused and ' not used' or '')
        io.write((" n: %d  key: %s%s\n"):format(bit.band(tree[index].cap, 0xffff) - 1, valuetable[tree[index].val], extra))
        printtree(tree, ident + 2, index + 1, valuetable);
        -- do not print next rule as a sibling
    elseif tag == TGrammar then
        local ruleindex = index + 1
        io.write((" %d\n"):format(tree[index].val)) -- number of rules
        for i = 1, tree[index].val do
            printtree(tree, ident + 2, ruleindex, valuetable);
            ruleindex = ruleindex + tree[ruleindex].ps
        end
        assert(tree[ruleindex].tag == TTrue); -- sentinel
    else
        local sibs = numsiblings[tree[index].tag] or 0
        io.write("\n")
        if sibs >= 1 then
            printtree(tree, ident + 2, index + 1, valuetable);
            if sibs >= 2 then
                printtree(tree, ident + 2, index + tree[index].ps, valuetable)
            end
        end
    end
end

-- }====================================================== */

return {
    printtree = printtree,
    printpatt = printpatt,
    printcaplist = printcaplist,
    printinst = printinst
}